File: leetcode/cpp/PassedProblems/211-add-and-search-word-data-structure-design.cpp
/*
 * 211-add-and-search-word-data-structure-design.cpp
 *
 *  Created on: Sep 17, 2015
 *      Author: hanzhe
 */

#include <iostream>
#include <map>
using namespace std;

class Node{
public:
	map<int,Node*> cn;
	bool isW;
	Node(){
		isW = false;
	}
};
class WordDictionary {
public:
	Node * root = new Node();

    // Adds a word into the data structure.
    void addWord(string word) {
    	if(root == NULL)
    		root = new Node();
    	Node * t = root;
        for(int i = 0 ; i < word.size(); i ++){
        	int ind =  word[i] - 'a';
        	if(t->cn.find(ind) == t->cn.end())
        		t->cn[ind] = new Node();
        	t = t->cn[ind];
        }
        t->isW = true;
    }

    // Returns if the word is in the data structure. A word could
    // contain the dot character '.' to represent any one letter.
    bool search(string word) {
    	return rec(word, root);
    }
    bool rec(string w, Node * r){
    	if(r == NULL)
    		return false;
    	if(w.length() == 0)
    		return r->isW;
    	int ind = w[0] - 'a';
//    	cout << "sub:" << w.substr(1) << endl;
    	if(w[0] == '.'){
    		for(int i = 0 ; i < 26; i ++)
    			if(r->cn.find(i) != r->cn.end() && rec(w.substr(1), r->cn[i]))
    				return true;
    	}
    	else if(r->cn.find(ind) != r->cn.end()){
    		return rec(w.substr(1), r->cn[ind]);
    	}
    	return false;
    }
};

class Solution{
public:
	void test(){
		WordDictionary d;
//		d.addWord("bad");
//		d.addWord("dad");
//		d.addWord("mad");
//		cout << d.search("pad");
//		cout << d.search("bad") ;
//		cout << d.search(".ad") ;
//		cout << d.search("b..") ;

		d.addWord("at");
		d.addWord("and");
		d.addWord("an");
		d.addWord("add");
		cout << "@" << endl;
		cout << d.search("a");
		cout << "@" << endl;
		cout << d.search(".at");
		d.addWord("bat");
		cout << "@" << endl;
		cout << d.search(".at");
		cout << d.search("an.");
		cout << d.search("a.d.");
		cout << d.search("b.");
		cout << d.search("a.d");
		cout << d.search(".");
	}
};
// Your WordDictionary object will be instantiated and called as such:
// WordDictionary wordDictionary;
// wordDictionary.addWord("word");
// wordDictionary.search("pattern");

File: leetcode/python/211-add-and-search-word-data-structure-design.py
#!/usr/bin/env python
# encoding: utf-8



"""

Design a data structure that supports the following two operations:

void addWord(word)
bool search(word)
search(word) can search a literal word or a regular expression string containing only letters a-z or .. A . means it can represent any one letter.

For example:

addWord("bad")
addWord("dad")
addWord("mad")
search("pad") -> false
search("bad") -> true
search(".ad") -> true
search("b..") -> true
Note:
You may assume that all words are consist of lowercase letters a-z.

"""


# 字符串查找、添加可以使用字典树Trie
# 4次过，忘写返回值，判断条件少node.end

class TrieNode(object):
    def __init__(self):
        self.val = None
        self.end = False
        self.child = {}

class WordDictionary(object):
    def __init__(self):
        """
        initialize your data structure here.
        """
        self.root = TrieNode()


    def addWord(self, word):
        """
        Adds a word into the data structure.
        :type word: str
        :rtype: void
        """
        t =self.root
        for x in word:
            if x not in t.child:
                t.child[x] = TrieNode()
                t = t.child[x]
                t.char = x
            else: t = t.child[x]
        t.end = True


    def search(self, word):
        """
        Returns if the word is in the data structure. A word could
        contain the dot character '.' to represent any one letter.
        :type word: str
        :rtype: bool
        """
        def inT(node, i):
            if i >= len(word): return True if node.end else False
            if word[i] in node.child:
                return inT(node.child[word[i]], i+1)
            elif word[i] == '.':
                for sn in node.child:
                    if inT(node.child[sn], i+1): return True
            return False
        return inT(self.root, 0)



# Your WordDictionary object will be instantiated and called as such:
# wordDictionary = WordDictionary()
# wordDictionary.addWord("word")
# wordDictionary.search("pattern")
